### ISSUE TITLE
GraphQLAuthorize Attribute

### ISSUE THREAD

**Issue by drowhunter** [ChilliCream/hotchocolate#451]
It would be really nice if we could add code like this in our underlying type

```
[GraphQLAuthorize("my policy")]
public Task<User> GetUser(IResolverContext context) {

}
```
instead of having to do 
```
protected override void Configure(IObjectTypeDescriptor<Query> descriptor)
{
            descriptor.Field(f => f.GetUser(default))
            .Authorize("my policy");
}
```

It would also be nice if this new attribute could be added to the Type itself
```
[GraphQLAuthorize("my policy")]
public class UserType : ObjectType<Query>{

}
```

or 

```
[GraphQLAuthorize("my policy")]
public class Query {   // <-- or you can add it to the underlying type instead

}
```

**Comment by drowhunter** [comment]
note I just realized that the authorize attribute also allows you to specify an array of roles
or both a policy and an array of roles

looking at the source code under abstractions i see that you have multiple extension methods for each place the attribute is added to and it returns string.  (for example with GetDescription)

but since authorize may need to return either a string policy or an array of roles would you need extension methods for each of them?

**Comment by drowhunter** [comment]
I was thinking the same thing but didnt think you would go for it :smile: .  Im noticing in alot of cases the only reason you have to use the Configure override is for things that you could do with attributes in the underlying type. Such as marking reference type parameters as NonNull , add directives to things..etc so it would be a nice feature.

**Comment by drowhunter** [comment]
only reason i was think about the authorize attribute is that it is similar to what we are already used to in normal mvc.  where you add `[Authorize]` to anything you want to protect. Its special i. that way from the other directives.

**Comment by jacob-ebey** [comment]
Is there any progress on this? I love being able to define my schema as models in their own project for client's to be able to consume, while the resolvers live in another project. The missing part to the architecture is an authorize attribute.

**Comment by michaelstaib** [comment]
@jacob-ebey sorry for the delay here. With version 9 we have put a lot of core apis in place to be able to add more attributes. So this one could be possible with 9.1.

**Comment by michaelstaib** [comment]
I will move it to the 9.1 backlog and it should be available in about 4 weeks.

**Comment by michaelstaib** [comment]
This one is implemented with 10.4

**Comment by drowhunter** [comment]
Just tried to add the new Authorize attribute by removing what I had which was
```
[Authorize(Policy="IsOver18")] //doesnt work here
public class AdultType:ObjectType<AdultResolver>
{
    protected override void Configure(IObjectTypeDescriptor<AdultResolver> descriptor){
   // removed this line but this works
   //  this descriptor.Authorize("IsOver18");

    }
}
[Authorize(Policy="IsOver18")] //this works
public class AdultResolver
{

}
```
unless im assuming too much on where the attribute should work.
